### PR TITLE
Fastify

### PR DESCRIPTION
**Pull request by tmcw**
Draft blog post about our switch to Fastify. This is kind of a narrow slice of it - just the decision process and some things about hono and express and history. I think this would fit in with a blog post that ships shortly thereafter about OpenAPI/stainless/typebox/etc - the SDK and type enforcement stack.

### CODE REVIEW

File: src/content/blog/fastify.mdx [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+});
+app.listen(3000);
+```
+
+Express is a beautifully light abstraction: a
+'hello world' example is barely over 10 lines. Compared to the boilerplate
+required to do the same in Ruby on Rails or a Java server at the time, it
+was a breath of fresh air.
+
+<NpmChart />
```
Comment (stevekrouse) [inline src/content/blog/fastify.mdx:40]: Interesting! Why this vs an image?

File: src/content/blog/fastify.mdx [hunk 2]
```diff
@@ -0,0 +1,179 @@
+---
+title: "Revamping our API pt. 1: Moving from express to Fastify"
```
Comment (stevekrouse) [inline src/content/blog/fastify.mdx:2]: The title is a bit wordy. How about:

- Migrating from Express to Fastify
- Migrating from ExpressJS - Part 1

File: src/content/blog/fastify.mdx [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+But the vibes have shifted: we write [TypeScript][ts] now,
+and everyone's excited about validator modules like [Zod](https://zod.dev/)
+and [TypeBox](https://github.com/sinclairzx81/typebox). Interfaces
+are less ad-hoc and more declared, tested, and strictly typed.
+
+Modern tooling is written to take full advantage of types
+and schemas. With [tRPC](https://trpc.io/), we're writing backend
+endpoints and calling them from auto-generated, auto-typed functions
+on the frontend. Other backends are embracing [OpenAPI][openapi]
+(née Swagger) to generate clients, test suites, and more.
```
Comment (stevekrouse) [inline src/content/blog/fastify.mdx:64]: I had to look up née lol. The joke I'd make is "the artist formerly known as Swagger"

File: src/content/blog/fastify.mdx [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+Val Town started off with Express. It has served us pretty well.
+
+But the papercuts started to accumulate.
+When we started creating  [a public API with an OpenAPI spec][rfc],
+we immediately started to get a bad feeling about using express
+as we were doing before. The OpenAPI specification was making
+strong promises about the kinds of requests we accepted and responses
+we'd return. But the specification was just a dead document,
+a YAML file - it wasn't part our codebase, or our test suite. So it
+wasn't enforced, and what isn't enforced isn't true.
```
Comment (stevekrouse) [inline src/content/blog/fastify.mdx:78]: You could also mention:

- our yaml file was never valid
- we wanted an sdk and that was hard (and you could link to https://macwright.com/2024/06/21/apis-and-applications)
- that openapi's are all the sudden more relevant in the age of llm's reading them for some reason

File: src/content/blog/fastify.mdx [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+  by the fastify folks, and it is built into fastify.
+
+Of course, fastify is also very fast, but web framework overhead
+wasn't a problem with express and most likely won't be a problem with
+any framework we choose: most of our performance envelope has to
+do with operating [the Val Town runtime][runtimes] and interacting
+with the database.
+
+We considered a few other options, like [Hono](https://hono.dev/) and [Elysia](https://elysiajs.com/).
+Both are lovely projects - we use and recommend Hono a ton
```
Comment (stevekrouse) [inline src/content/blog/fastify.mdx:142]: Probably a good place to do a shoutout to Hono and maybe even link out to a val that uses it. If this article has a "go use val town" component, I think this may be it

File: src/content/blog/fastify.mdx [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+incremental migration from express, and plugins - was more built-out.
+
+For simpler projects and especially for situations where we want
+multi-platform compatibility so that code can run on Node.js, Deno,
+or Cloudflare Workers, we reach for Hono.
+
+### The port
+
+By using [@fastify/express](https://github.com/fastify/fastify-express)
+we were able to do this migration entirely incrementally: the first
```
Comment (stevekrouse) [inline src/content/blog/fastify.mdx:155]: In some ways I feel like this is the most interesting part of the post and I wonder if we can get it moved up higher or even in the title, ie "Migrating from Express to Fastify, one route a time"

File: src/content/blog/fastify.mdx [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+server with fastify. Then each successive PR moved a few routes
+from express to fastify, until the application has been
+completely ported.
+
+Of course there were some bumps along the road, and lots of details
+to iron out. But as of today we're nearly 100% fastify for Val Town's
+backend, and it's unlocking stability and capabilities across
+the board. Stay tuned for the next update which will dive into those,
+including how this new backend strategy lets us automatically
+generate our new TypeScript SDK!
```
Comment (stevekrouse) [inline src/content/blog/fastify.mdx:166]: Link to the SDK on github?

File: src/content/blog/fastify.mdx [hunk 8]
```diff
@@ -0,0 +1,179 @@
...
+
+[openapi]: https://en.wikipedia.org/wiki/OpenAPI_Specification
```
Comment (stevekrouse) [inline src/content/blog/fastify.mdx:168]: Didn't know about this feature of markdown or mdx!

**Comment by stevekrouse** [comment]
Re-reading all my suggestions - they're all skippable in my opinion. The only one I really care about is the title

**Comment by tmcw** [comment]
Thanks for the comments! Agree with a lot of them, I'll probably do a big editing pass tomorrow

File: src/content/blog/fastify.mdx (hunk 1 above)
Comment (tmcw) [reply src/content/blog/fastify.mdx:40]: Primarily because I want to to look good on both mobile and desktop, and this does that nicely instead of making text too tiny. And besides that, will probably have charts in future blog posts and wanted to get a handle on what the patterns were.

File: src/content/blog/fastify.mdx (hunk 8 above)
Comment (tmcw) [reply src/content/blog/fastify.mdx:168]: Yep, they're called reference-style links - https://www.markdownguide.org/basic-syntax/#reference-style-links
